cdap/resource_streaming_program_run.go

```diff
@@ -0,0 +1,335 @@
+	if err != nil {
+		return
+	}
+
+	b, err := httpCall(config.httpClient, req)
+	if err != nil {
+		return
+	}
+
+	var runs []Run
+	var args RuntimeArgs
+
+	err = json.Unmarshal(b, &runs)
+	if err != nil {
+		return
+	}
+
+	for _, r := range runs {
+		// Unescaping runtime Args which are stored as an escaped JSON string.
+		unquotedArgs, _ := strconv.Unquote(string(r.Properties.RuntimeArgs))
+		log.Println(fmt.Sprintf("%s", unquotedArgs))
+		b := []byte(unquotedArgs)
+		if err := json.Unmarshal(b, &args); err != nil {
+			// This happens when a run does not contain the special TF con
+			log.Println("failed to parse unquotedArgs json")
+		}
+		log.Println(fmt.Sprintf("found terraform faux run id: %s", args.FauxRunId))
+		log.Println(fmt.Sprintf("status: %s", r.Status))
+		if runId == args.FauxRunId {
+			return r, nil
+		}
+	}
+	err = fmt.Errorf("no run found with faux runid: %s", runId)
+	return
+}
+
+func stopProgramRun(config *Config, stopAddr string) error {
+	req, err := http.NewRequest(http.MethodPost, stopAddr, nil)
+	if err != nil {
+		return err
+	}
+	_, err = httpCall(config.httpClient, req)
+	return err
+}
+
+func resourceStreamingProgramRunDelete(d *schema.ResourceData, m interface{}) error {
+	config := m.(*Config)
+
+	addr := getProgramAddr(config, d)
+	runsAddr := urlJoin(addr, "/runs")
+	r, err := getRunByFauxId(config, runsAddr, d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	stopAddr := urlJoin(runsAddr, r.RunId, "/stop")
+
+	return resource.Retry(d.Timeout(schema.TimeoutDelete), func() *resource.RetryError {
+		s, err := getProgramRunStatusByFauxId(config, runsAddr, d.Id())
+
+		if s == "RUNNING" || INITIALIZING_PROGRAM_RUN_STATUSES[s] {
+			err = stopProgramRun(config, stopAddr)
+			if err != nil {
+				return resource.NonRetryableError(fmt.Errorf("error stopping program: %s", err))
+			}
+		}
+
+		if END_PROGRAM_STATUSES[s] {
+			return nil
+		}
+
+		return resource.NonRetryableError(fmt.Errorf("failed to delete run with id: %s and faux id: %s in status: %s", r.RunId, d.Id(), s))
+	})
+}
+
+func resourceStreamingProgramRunExists(d *schema.ResourceData, m interface{}) (bool, error) {
+	config := m.(*Config)
+
+	addr := getProgramAddr(config, d)
+	statusAddr := urlJoin(addr, "/status")
+	runAddr := urlJoin(addr, "/runs")
+
+	req, err := http.NewRequest(http.MethodGet, statusAddr, nil)
+	if err != nil {
+		return false, err
+	}
+
+	b, err := httpCall(config.httpClient, req)
+	if err != nil {
+		return false, err
+	}
+
+	var p ProgramStatus
+	if err := json.Unmarshal(b, &p); err != nil {
+		return false, err
+	}
+
+	// This checks if there program is running (but it may be running several times)
+	if p.Status == "RUNNING" {
+		// This handles ambiguity if there are multiple program runs
+		running, err := isFauxRunIdRunningYet(config, runAddr, d.Id())
+		if err != nil { // We
+			return false, nil
+		}
+
+		if running {
+			return true, nil
+		}
+
+		if d.Get("allow_multiple_runs").(bool) {
```

Comment:
Exists should not care about other resources. The reason is exists is only called AFTER create, so I can actually create multiple runs and then only on future applies would they all suddenly start failing.

Let's just remove the allow multiple check for v1 and we can add this in the future if it's needed. If you think this is an essential feature this probably needs to be done in the create.